File: clients/go-client/solidb/roles.go
package solidb

// RolesClient provides access to RBAC roles management API
type RolesClient struct {
	client *Client
}

// Permission represents a role permission
type Permission struct {
	Action   string  `msgpack:"action"`
	Scope    string  `msgpack:"scope"`
	Database *string `msgpack:"database,omitempty"`
}

// List returns all roles
func (r *RolesClient) List() ([]interface{}, error) {
	res, err := r.client.SendCommand("list_roles", nil)
	if err != nil {
		return nil, err
	}
	if slice, ok := res.([]interface{}); ok {
		return slice, nil
	}
	return nil, nil
}

// Create creates a new role
func (r *RolesClient) Create(name string, permissions []map[string]interface{}, description *string) (map[string]interface{}, error) {
	args := map[string]interface{}{
		"name":        name,
		"permissions": permissions,
	}
	if description != nil {
		args["description"] = *description
	}
	res, err := r.client.SendCommand("create_role", args)
	if err != nil {
		return nil, err
	}
	if m, ok := res.(map[string]interface{}); ok {
		return m, nil
	}
	return nil, nil
}

// Get retrieves a role by name
func (r *RolesClient) Get(name string) (map[string]interface{}, error) {
	res, err := r.client.SendCommand("get_role", map[string]interface{}{
		"role_name": name,
	})
	if err != nil {
		return nil, err
	}
	if m, ok := res.(map[string]interface{}); ok {
		return m, nil
	}
	return nil, nil
}

// Update modifies an existing role
func (r *RolesClient) Update(name string, permissions []map[string]interface{}, description *string) (map[string]interface{}, error) {
	args := map[string]interface{}{
		"role_name":   name,
		"permissions": permissions,
	}
	if description != nil {
		args["description"] = *description
	}
	res, err := r.client.SendCommand("update_role", args)
	if err != nil {
		return nil, err
	}
	if m, ok := res.(map[string]interface{}); ok {
		return m, nil
	}
	return nil, nil
}

// Delete removes a role
func (r *RolesClient) Delete(name string) error {
	_, err := r.client.SendCommand("delete_role", map[string]interface{}{
		"role_name": name,
	})
	return err
}

File: clients/elixir_client/lib/solidb/roles.ex
defmodule SoliDB.Roles do
  @moduledoc """
  RBAC roles management API for SoliDB.
  """

  alias SoliDB.Client

  def list(client) do
    Client.send_command(client, "list_roles", %{})
  end

  def create(client, name, permissions, opts \\ []) do
    args = %{name: name, permissions: permissions}
    args = if opts[:description], do: Map.put(args, :description, opts[:description]), else: args
    Client.send_command(client, "create_role", args)
  end

  def get(client, name) do
    Client.send_command(client, "get_role", %{role_name: name})
  end

  def update(client, name, permissions, opts \\ []) do
    args = %{role_name: name, permissions: permissions}
    args = if opts[:description], do: Map.put(args, :description, opts[:description]), else: args
    Client.send_command(client, "update_role", args)
  end

  def delete(client, name) do
    Client.send_command(client, "delete_role", %{role_name: name})
  end
end
